crypto/asn1_dsa.c

```diff
@@ -0,0 +1,235 @@
+ * (i.e. rounded down), plus one.
+ * If n is zero then the content is a single zero byte (length = 1).
+ * If the number of bits of n is a multiple of 8 then an extra zero padding
+ * byte is included to ensure that the value is still treated as positive in
+ * the INTEGER two's complement representation.
+ */
+#define integer_content_len(n) (BN_num_bits(n) / 8 + 1)
+
+size_t calc_der_length(size_t cont_len)
+{
+    size_t llen;
+
+    if (cont_len <= 0x7f)
+        llen = 1;
+    else if (cont_len <= 0xff)
+        llen = 2;
+    else if (cont_len <= 0xffff)
+        llen = 3;
+    else
+        return 0;
+    return 1 + llen + cont_len;
+}
+
+/*
+ * Outputs the encoding of the length octets for a DER value with a content
+ * length of cont_len to *ppout and increments *ppout past the data just
+ * written.
+ */
+void encode_der_length(size_t cont_len, unsigned char **ppout)
+{
+    if (cont_len > 0xff) {
+        *(*ppout)++ = 0x82;
+        *(*ppout)++ = (unsigned char)(cont_len >> 8);
+    } else if (cont_len > 0x7f) {
+        *(*ppout)++ = 0x81;
+    }
+    *(*ppout)++ = (unsigned char)cont_len;
+}
+
+/*
+ * Calculates and returns the size of the DER encoding of an INTEGER with the
+ * value n.
+ */
+size_t calc_integer_der_length(const BIGNUM *n)
+{
+    if (BN_is_negative(n))
+        return 0;
+    return calc_der_length(integer_content_len(n));
+}
+
+void encode_der_integer(const BIGNUM *n, unsigned char **ppout)
+{
+    size_t len = integer_content_len(n);
+    int outlen;
+
+    OPENSSL_assert(!BN_is_negative(n));
+    *(*ppout)++ = ID_INTEGER;
+    encode_der_length(len, ppout);
+    outlen = BN_bn2binpad(n, *ppout, (int)len);
+    OPENSSL_assert(outlen == (int)len);
+    *ppout += len;
+}
+
+/*
+ * Outputs the DER encoding of a DSA-Sig-Value or ECDSA-Sig-Value to *ppout
+ * and increments *ppout past the data just written.
+ * If either ppout or *ppout is NULL then the output size is calculated and
+ * returned but no output is produced.
+ * Returns the number of bytes of output produced (or that would be produced)
+ * or 0 if an error occurs.
+ */
+size_t encode_der_dsa_sig(const BIGNUM *r, const BIGNUM *s,
+                          unsigned char **ppout)
+{
+    size_t r_der_len;
+    size_t s_der_len;
+    size_t seq_cont_len;
+    size_t seq_enc_len;
+
+    if ((r_der_len = calc_integer_der_length(r)) == 0
+            || (s_der_len = calc_integer_der_length(s)) == 0)
+        return 0;
+
+    seq_cont_len = r_der_len + s_der_len;
+    if ((seq_enc_len = calc_der_length(seq_cont_len)) == 0)
+        return 0;
+
+    if (ppout != NULL && *ppout != NULL) {
+        *(*ppout)++ = ID_SEQUENCE;
+        encode_der_length(seq_cont_len, ppout);
+        encode_der_integer(r, ppout);
+        encode_der_integer(s, ppout);
+    }
+    return seq_enc_len;
+}
+
+/*
+ * Decodes the DER length octets at *ppin, stores the decoded length to
+ * *pcont_len and increments *ppin past the data that was consumed.
```

Comment:
Might I suggest that `**ppin` only gets updated on successful decode, i.e. when this function returns 0, it should be left untouched?

Reply:
This goes for all decoding functions, of course

Reply:
Yeah, okay. Sure.

Reply:
I have changed all the decoding functions so that `*ppin` is only advanced on successful decode.
After adding bounds checking to the encoding functions I also did the same thing for `*ppout`.